server/handlers/validation.go

```diff
@@ -0,0 +1,377 @@
+package handlers
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/endophage/gotuf"
+	"github.com/endophage/gotuf/data"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/notary/server/storage"
+	"github.com/endophage/gotuf/keys"
+	"github.com/endophage/gotuf/signed"
+	"github.com/endophage/gotuf/utils"
+)
+
+// ErrValidation represents a general validation error
+type ErrValidation struct {
+	msg string
+}
+
+func (err ErrValidation) Error() string {
+	return fmt.Sprintf("An error occurred during validation: %s", err.msg)
+}
+
+// ErrBadHeirarchy represents a missing snapshot at this current time.
+// When delegations are implemented it will also represent a missing
+// delegation parent
+type ErrBadHeirarchy struct {
+	msg string
+}
+
+func (err ErrBadHeirarchy) Error() string {
+	return fmt.Sprintf("Heirarchy of updates in incorrect: %s", err.msg)
+}
+
+// ErrBadRoot represents a failure validating the root
+type ErrBadRoot struct {
+	msg string
+}
+
+func (err ErrBadRoot) Error() string {
+	return fmt.Sprintf("The root being updated is invalid: %s", err.msg)
+}
+
+// ErrBadTargets represents a failure to validate a targets (incl delegations)
+type ErrBadTargets struct {
+	msg string
+}
+
+func (err ErrBadTargets) Error() string {
+	return fmt.Sprintf("The targets being updated is invalid: %s", err.msg)
+}
+
+// ErrBadSnapshot represents a failure to validate the snapshot
+type ErrBadSnapshot struct {
+	msg string
+}
+
+func (err ErrBadSnapshot) Error() string {
+	return fmt.Sprintf("The snapshot being updated is invalid: %s", err.msg)
+}
+
+// validateUpload checks that the updates being pushed
+// are semantically correct and the signatures are correct
+func validateUpdate(gun string, updates []storage.MetaUpdate, store storage.MetaStore) error {
+	kdb := keys.NewDB()
+	repo := tuf.NewTufRepo(kdb, nil)
+	rootRole := data.RoleName(data.CanonicalRootRole)
+	targetsRole := data.RoleName(data.CanonicalTargetsRole)
+	snapshotRole := data.RoleName(data.CanonicalSnapshotRole)
+
+	// check that the necessary roles are present:
+	roles := make(map[string]storage.MetaUpdate)
+	for _, v := range updates {
+		roles[v.Role] = v
+	}
+	if err := heirarchyOK(roles); err != nil {
+		logrus.Error("ErrBadHeirarchy: ", err.Error())
```

Comment:
I suggest passing in a context to this function and using ctxu.GetLogger().Error instead of logrus.Error. Then this log line will indicate which request experienced the error. See https://github.com/docker/notary/pull/174